### ISSUE TITLE
Docs: add Desktop → Training page for the new Train tab (PR #4372)

### ISSUE THREAD

**Issue by MervinPraison** [MervinPraison/PraisonAIDocs#3366]
## Summary

MervinPraison/PraisonAI PR [#4372](https://github.com/MervinPraison/PraisonAI/pull/4372) (`feat(desktop): add a Training tab that runs a fine-tune and reports on it`, merged 2026-08-25) shipped a brand-new **Train** tab in the PraisonAI Desktop app. There is no page for it in `docs/features/desktop/` yet, and the rest of the desktop docs still describe Chat as the only workspace. This issue tracks the docs work.

## Decision: new content (not an update)

The Desktop docs (`docs/features/desktop/{index,chat,mcp,models,settings,approvals,conversations,data,troubleshooting}.mdx`) do not mention training at all. The existing `praisonai-train-*` pages under `docs/features/` (and `docs/train.mdx`, `docs/cli/train.mdx`) document the **CLI** — this new feature is a **desktop UI wrapper** around that CLI, so it needs its own page, sibling to `chat.mdx`.

**Primary deliverable:** create a new `docs/features/desktop/training.mdx`.

**Secondary updates:**
- `docs/features/desktop/index.mdx` — add a Train card in the "What's in the app" grid and mention Train in the overview sentence
- `docs.json` (Mintlify navigation) — add the new page under the Desktop group, right after Chat
- `docs/train.mdx` and `docs/features/praisonai-train-package.mdx` — one-liner cross-link Card pointing at the new desktop page ("Prefer a UI? See Desktop → Training")

## Placement rules

Per `AGENTS.md`: place the new page in `docs/features/desktop/` (`docs/features/` is AI-agent-writable). Do **not** touch `docs/concepts/`.

---

## Source-of-truth files (read these before writing)

All paths are in `MervinPraison/PraisonAI` at commit `ec3a260c9a052242d18aaf518c659bbe38aab5a9` (PR #4372 head, now merged to `main`):

| File | Why |
|---|---|
| `src/praisonai-desktop/engine/training.py` (new, +420) | `Trainer`, `Run`, `parse_metric`, `parse_progress`, `validate_run_id`, `split_command`, ring buffer, YAML config write, process-group termination. Truth for **every config field, event kind, and status value**. |
| `src/praisonai-desktop/engine/server.py` (modified, +111) | Adds the five `/train/*` HTTP routes and the SSE progress stream (heartbeat + `resync` event). Truth for **endpoint paths, request/response bodies, status codes**. |
| `src/praisonai-desktop/ui/index.html` (modified, +462) | The Train tab UI: form, live card, loss chart, log tail, history list, method hints, tab switching, `refreshTraining()` adopt-in-flight, resync message. Truth for **user-facing labels, defaults, and behaviors**. |
| `src/praisonai-desktop/engine/test_training.py` (new, +467) | Behavioral guarantees under test. |
| `src/praisonai-desktop/engine/test_train_routes.py` (new, +232) | End-to-end HTTP contracts under test. |
| `src/praisonai-desktop/frontend/tests/training.test.mjs` (new, +313) | UI contracts under test. |

For the wrapped CLI, cross-reference the existing docs (do not re-document): `docs/train.mdx`, `docs/features/praisonai-train-package.mdx`, `docs/features/praisonai-train-preference-tuning.mdx`, `docs/features/praisonai-train-cpt.mdx`.

---

## What the feature does (for the doc writer)

The Desktop app previously talked *to* a model but could not *make* one. The Train tab lets a user configure a local fine-tune, kick it off, watch a live loss chart + log, stop it, and see history — all against the local engine, which spawns `python -m praisonai_train llm --config <run_dir>/config.yaml` as its own process group.

**Design constraints worth surfacing (why the UI looks like it does):**
- **One run at a time.** Second `POST /train/start` while a run is live returns `409` naming the blocking run id. No queue.
- **Progress is a replayable ring buffer, not a stream.** Reconnecting reads from a cursor; if history was evicted, engine emits `event: resync` and the UI appends "… earlier output dropped; the full log is in the run directory." then resets cursor to `-1`. Closing the lid loses nothing.
- **Adopt-in-flight.** Opening the Train tab in a fresh window reattaches to a run started in a previous session and repopulates step / loss / elapsed / log.
- **Stop is a `SIGTERM` to the child's process group** (POSIX) or `CTRL_BREAK_EVENT` (Windows), so dataloader workers and torchrun ranks die with it. The engine itself is protected (checks child pgid differs from own).
- Full log is persisted to `<PRAISONAI_DESKTOP_HOME>/runs/<run_id>/train.log` (UTF-8, `errors="replace"`); config is written next to it as `config.yaml` (falls back to JSON if PyYAML is missing).

---

## HTTP endpoints (document under a Reference section)

| Method | Path | Purpose | Notable status codes |
|---|---|---|---|
| `POST` | `/train/start` | Start a new run. Body: `{config, run_id?}`. | `200` run summary; `400` when `model_name` or `dataset` missing or `run_id` invalid; `409` when a run is already live |
| `POST` | `/train/stop` | Stop the live run. | `200` on stop; `404` if nothing running |
| `POST` | `/train/stop/<run_id>` | Stop a specific run. | `409` if `run_id` is not the live run |
| `GET`  | `/train/status` | Current run summary + last 500 metric samples. | `200` `{run: summary\|null, metrics: [...]}` |
| `GET`  | `/train/runs` | Last 50 run summaries. | `200` `{runs: [...]}` |
| `GET`  | `/train/progress?run=<id>&cursor=<n>` | SSE stream. Replays from cursor then follows. Keepalive comment every ~5s. | `404` if run unknown |

**SSE event kinds:** `start`, `state`, `log`, `progress`, `metric`, `end`, plus `resync` when history was evicted past the client's cursor.

**Run summary shape:** `{id, state, step, total, started, ended, error, elapsed, last_loss}`. `state` is one of `running`, `done`, `failed`, `cancelled`, `stopping`.

---

## Training config fields (posted as `config` on `/train/start`)

Extract these verbatim from `training.py` and the `trainConfig()` JS in `ui/index.html`:

| Field | Type | Default | Notes |
|---|---|---|---|
| `model_name` | str | `unsloth/Meta-Llama-3.1-8B-Instruct-bnb-4bit` | Required |
| `method` | enum | `sft` | `sft`, `cpt`, `dpo`, `orpo`, `kto`, `grpo`, `reward`, `cpo` |
| `dataset` | list | `[{name: "yahma/alpaca-cleaned", split_type: "train"}]` | Required; sent as a list even when the user picked one name |
| `split_type` | str | `train` | Folded into `dataset[0].split_type` before posting |
| `num_train_epochs` | float | `1` | |
| `max_steps` | int | `0` | **Omitted from payload when `0`** — 0 means "no cap" |
| `lora_r` | int | `16` | |
| `lora_alpha` | int | `16` | |
| `learning_rate` | float | `2e-4` | |
| `max_seq_length` | int | `2048` | |
| `per_device_train_batch_size` | int | `2` | |
| `gradient_accumulation_steps` | int | `2` | |
| `load_in_4bit` | bool | `true` | |
| `output_dir` | str | `outputs` | |

Also fixed by the engine before writing `config.yaml`: `train: true`, `ollama_save: false`, `huggingface_save: false`.

---

## Environment variables

| Var | Purpose |
|---|---|
| `PRAISONAI_TRAIN_CMD` | Overrides the trainer launcher. Parsed with `split_command` (POSIX rules on POSIX, Windows rules on Windows). `--config <path>` is always appended. Default when unset: `[<python>, "-m", "praisonai_train", "llm"]`. Use this to point the desktop app at a separate CUDA-matched environment. |
| `PRAISONAI_DESKTOP_HOME` | Existing var; runs live under `<home>/runs/<run_id>/{config.yaml, train.log}`. Already documented in `docs/features/desktop/data.mdx` — cross-link. |

---

## UI features (for the "Quick Start" and "How it works" sections)

- Sidebar tabs (`Chat` / `Train`); last-picked view is remembered in `localStorage['view']`.
- Basic form + collapsible **Advanced** panel (LoRA, batch, quantization, output).
- `Start training` button → becomes `Training…`. Second click while running triggers the 409 path; UI shows the engine's message and re-enables the button; the live run is not disturbed.
- Live card: progress bar, `step / total`, current loss, elapsed, method hint (swapped per `method` via `METHOD_HINTS`).
- Loss chart (`<canvas id="loss">`) drawn from `metric` events, DPR-scaled, endpoint dot at latest sample.
- Log pane: tail of 600 lines (`MAX_LOG_LINES`), auto-follows only when the reader is already pinned to the bottom.
- Status pill classes: `running` / `done` / `failed` / `cancelled` / `stopping`.
- Error paragraph (`role="alert"`) surfaces the engine's failure message.
- History card: last 50 runs with id, pill, `last_loss`, elapsed, and failure reason in the `title` tooltip.

---

## Recommended page structure

Follow `AGENTS.md` (hero Mermaid → Quick Start Steps → How It Works → Configuration → Common Patterns → Best Practices → Related). Non-developer, user-centric, and *agent-first* framing per the doc principles.

Suggested frontmatter:
```yaml
---
title: "Training"
sidebarTitle: "Training"
description: "Fine-tune a local model from the Desktop app — one run at a time, with live loss and reconnect-safe progress"
icon: "graduation-cap"
---
```

Suggested Mermaid hero (using the standard palette):
```mermaid
graph LR
    subgraph "Desktop Training"
        F[📝 Form] --> S[🚀 Start]
        S --> E[⚙️ Engine spawns praisonai-train]
        E --> P[📈 Live loss + log]
        P --> D[💾 outputs/ & run history]
    end
    classDef input fill:#6366F1,stroke:#7C90A0,color:#fff
    classDef process fill:#F59E0B,stroke:#7C90A0,color:#fff
    classDef result fill:#10B981,stroke:#7C90A0,color:#fff
    class F input
    class S,E,P process
    class D result
```

A sequence diagram for the reconnect/resync flow will help — cover: client opens SSE with `cursor=N` → engine replays from ring buffer → if `N` was evicted, engine sends `event: resync` → UI appends the "earlier output dropped" line and resets cursor.

An option-selection diagram for `method` (SFT vs preference tuning vs CPT vs GRPO) — link each branch to the existing `praisonai-train-*` page.

Suggested Quick Start Steps:
1. **Open the Train tab.** Click Train in the sidebar.
2. **Pick a model and dataset.** Defaults are `unsloth/Meta-Llama-3.1-8B-Instruct-bnb-4bit` on `yahma/alpaca-cleaned`.
3. **Click Start training.** Live loss appears; log auto-follows.
4. **Close the lid — come back.** Reopen the app; the Train tab reattaches to the running job.

Suggested Related cards:
- `Chat & Streaming` → `/features/desktop/chat`
- `Data & Privacy` → `/features/desktop/data`
- `Train (CLI)` → `/train`
- `Preference Tuning` → `/features/praisonai-train-preference-tuning`

---

## User-interaction scenarios to include

1. **First fine-tune.** User launches the app → clicks Train → keeps defaults → clicks Start. Watches loss drop. Model saved to `outputs/`.
2. **Interrupted run.** Laptop lid closes mid-training. User opens app hours later; the Train tab reattaches, loss chart repopulates from history, run finishes.
3. **Second start while one is running.** User forgets a run is going; clicks Start again; UI surfaces the engine's 409 message; the live run continues undisturbed.
4. **Stop.** User clicks Stop; the run and all its dataloader workers die together; status pill turns `cancelled`; log tail preserved.
5. **Method hint swap.** User picks `dpo` from the method dropdown; the hint under it changes to explain the different dataset shape required (`prompt` / `chosen` / `rejected`); cross-link the CLI page for the full shape.
6. **Separate CUDA env.** Power user exports `PRAISONAI_TRAIN_CMD="/opt/cuda-venv/bin/python -m praisonai_train llm"` before launching the app so the engine uses a matched CUDA/torch environment.

---

## Do / Don't checklist for the writing agent

- [ ] Read the six source files above **before** drafting a single paragraph.
- [ ] Copy defaults and field names **verbatim** from `training.py` — don't paraphrase types.
- [ ] Keep it agent-centric (per AGENTS.md §1.1): open with a code/agent-shaped example, not a UI screenshot description.
- [ ] Show the config table in full — use progressive disclosure (basic fields first, LoRA/quantization in an `<AccordionGroup>`).
- [ ] Use the standard Mermaid palette (`#8B0000`, `#189AB4`, `#10B981`, `#F59E0B`, `#6366F1`, white text, `#7C90A0` strokes).
- [ ] Cross-link — don't duplicate — the CLI training pages. This page describes the *desktop wrapper*, not `praisonai-train llm` itself.
- [ ] Update `docs/features/desktop/index.mdx` to reference Training in the overview and add a Card in the grid.
- [ ] Update `docs.json` to slot Training right after Chat under the Desktop group.
- [ ] Do **not** create anything under `docs/concepts/`.

## References

- Source PR: https://github.com/MervinPraison/PraisonAI/pull/4372
- Merged commit: `ec3a260c9a052242d18aaf518c659bbe38aab5a9`
- Existing desktop docs pattern: `docs/features/desktop/chat.mdx`
- Existing training CLI docs (for cross-linking): `docs/train.mdx`, `docs/features/praisonai-train-*.mdx`
